src/ucp/wireup/wireup_cm.c

```diff
@@ -245,24 +245,75 @@ ucp_cm_client_connect_prog_arg_free(ucp_cm_client_connect_progress_arg_t *arg)
     ucs_free(arg);
 }
 
-static void ucp_cm_client_restore_ep(ucp_wireup_ep_t *wireup_cm_ep,
-                                     ucp_ep_h ucp_ep)
+/**
+ * Copies lanes from the one UCP EP to the another UCP EP. The function
+ * creates new WIREUP EPs for all lanes in @to_ep and sets UCT EP of
+ * the TLs from @from_ep. Both EPs have to be created and initalized.
+ * @to_ep should not have initialized the lanes by UCT EPs that will be
+ * overwritten by UCT EPs from the @from_ep's lanes.
+ * After copying UCT EPs, @to_ep should try to reconfigure lanes and
+ * some of the copied UCT EPs could be re-used in the new configuration.
+ *
+ * @param [in] to_ep               UCP EP handle to copy the lanes to.
+ * @param [in] from_ep             UCP EP handle to copy the lanes from.
+ * @param [in] change_ownership    Make WIREUP EPs in @to_ep are owner for
+ *                                 copied UCT EPs from @from_ep.
+ */
+static void ucp_cm_copy_ep_lanes(ucp_ep_h to_ep, ucp_ep_h from_ep,
+                                 int change_ownership)
 {
-    ucp_ep_h tmp_ep = wireup_cm_ep->tmp_ep;
-    ucp_wireup_ep_t *w_ep;
+    int to_is_owner, from_is_owner;
     ucp_lane_index_t lane_idx;
+    ucs_status_t status;
+    uct_ep_h uct_ep;
+    int is_local_connected;
+    ucp_wireup_ep_t *from_wireup_ep;
 
-    for (lane_idx = 0; lane_idx < ucp_ep_num_lanes(tmp_ep); ++lane_idx) {
-        if (tmp_ep->uct_eps[lane_idx] != NULL) {
-            ucs_assert(ucp_ep->uct_eps[lane_idx] == NULL);
-            ucp_ep->uct_eps[lane_idx] = tmp_ep->uct_eps[lane_idx];
-            w_ep = ucs_derived_of(ucp_ep->uct_eps[lane_idx], ucp_wireup_ep_t);
-            w_ep->super.ucp_ep = ucp_ep;
+    to_is_owner   = change_ownership;
+    from_is_owner = !change_ownership;
+
+    for (lane_idx = 0; lane_idx < ucp_ep_num_lanes(from_ep); ++lane_idx) {
+        if ((lane_idx == ucp_ep_get_cm_lane(from_ep)) ||
+            (from_ep->uct_eps[lane_idx] == NULL)) {
+            continue;
+        }
+
+        ucs_assert_always(to_ep->uct_eps[lane_idx] == NULL);
+
+        uct_ep = ucp_wireup_extract_lane(from_ep, lane_idx);
+        if (uct_ep == NULL) {
+            /* UCT EP could be NULL only for non-P2P TLs */
+            ucs_assert(ucp_worker_is_tl_2iface(from_ep->worker,
+                                               ucp_ep_config(from_ep)->key.
+                                               lanes[lane_idx].rsc_index));
+            continue;
         }
-    }
 
-    ucp_ep_destroy_base(tmp_ep); /* not needed anymore */
-    wireup_cm_ep->tmp_ep = NULL;
+        from_wireup_ep = ucp_wireup_ep(from_ep->uct_eps[lane_idx]);
+
+        is_local_connected = (from_wireup_ep == NULL) ||
+                             (from_wireup_ep->flags &
+                              UCP_WIREUP_EP_FLAG_LOCAL_CONNECTED);
+
+        status = ucp_wireup_ep_create(to_ep, &to_ep->uct_eps[lane_idx]);
```

Comment:
can we just take existing wireup ep, like it was done before?

Reply:
no, since we will need this to set who is the owner of the UCT EP- TMP EP or user's UCP EP